src/control/server/storage_scm.go

```diff
@@ -57,17 +114,194 @@ var (
 type scmStorage struct {
 	ipmctl      ipmctl.IpmCtl  // ipmctl NVM API interface
 	config      *configuration // server configuration structure
+	runCmd      runCmdFn
 	modules     common.ScmModules
+	pmemDevs    []pmemDev
+	state       scmState
 	initialized bool
 	formatted   bool
 }
 
+func (s *scmStorage) withRunCmd(runCmd runCmdFn) *scmStorage {
+	s.runCmd = runCmd
+
+	return s
+}
+
 // TODO: implement remaining methods for scmStorage
 // func (s *scmStorage) Update(req interface{}) interface{} {return nil}
 // func (s *scmStorage) BurnIn(req interface{}) (fioPath string, cmds []string, env string, err error) {
 // return
 // }
 
+// Prep executes commands to configure SCM modules into AppDirect interleaved
+// regions/sets hosting pmem kernel device namespaces.
+//
+// Presents of nonvolatile memory modules is assumed in this method and state
+// is established based on presence and free capacity of regions.
+//
+// Actions based on state:
+// * modules exist and no regions -> create all regions (needs reboot)
+// * regions exist and free capacity -> create all namespaces
+// * regions exist but no free capacity -> no-op
+//
+// Command output from external tools will be returned.
+func (s *scmStorage) Prep() (needsReboot bool, pmemDevs []pmemDev, err error) {
+	if err := s.getState(); err != nil {
+		return false, nil, errors.WithMessage(err, "establish scm state")
+	}
+
+	log.Debugf("scm in state %s\n", s.state)
+
+	switch s.state {
+	case scmStateNoRegions:
+		needsReboot, err = s.createRegions()
+	case scmStateFreeCapacity:
+		pmemDevs, err = s.createNamespaces()
+	case scmStateNoCapacity:
+		pmemDevs, err = s.getNamespaces()
+	default:
+		err = errors.New("unknown scm state")
+	}
+
+	return
+}
+
+// reset executes commands to remove namespaces and regions on SCM models.
+func (s *scmStorage) PrepReset() error {
+	return nil // TODO
+}
+
+// getState establishes state of SCM regions and namespaces on local server.
+func (s *scmStorage) getState() error {
+	s.state = scmStateUnknown
+
+	// TODO: discovery should provide SCM region details
+	out, err := s.runCmd(cmdScmShowRegions)
+	if err != nil {
+		return err
+	}
+
+	if out == outScmNoRegions {
+		s.state = scmStateNoRegions
+		return nil
+	}
+
+	ok, err := hasFreeCapacity(out)
+	if err != nil {
+		return err
+	}
+	if ok {
+		s.state = scmStateFreeCapacity
+		return nil
+	}
+	s.state = scmStateNoCapacity
+
+	return nil
+}
+
+// hasFreeCapacity takes output from ipmctl and checks for free capacity.
+//
+// external tool commands return:
+// $ ipmctl show -d PersistentMemoryType,FreeCapacity -region
+//
+// ---ISetID=0x2aba7f4828ef2ccc---
+//    PersistentMemoryType=AppDirect
+//    FreeCapacity=3012.0 GiB
+// ---ISetID=0x81187f4881f02ccc---
+//    PersistentMemoryType=AppDirect
+//    FreeCapacity=3012.0 GiB
+//
+// FIXME: implementation to be replaced by using libipmctl directly through bindings
```

Comment:
Even in the short term it would be better to use the XML-formatted output from ipmctl if possible.

Reply:
I think switching to bindings for region discovery should be trivial

Reply:
work scheduled in https://jira.hpdd.intel.com/browse/DAOS-2758

Reply:
OK with it as long as we move to the bindings sooner rather than later.